### PR TITLE
iQMC Refactor

### PR DESCRIPTION
**Pull request by spasmann**
This PR includes two significant changes:
1) All of the iQMC functions in both `loop.py` and `kernel.py` have been moved into `iqmc_loop.py` and `iqmc_kernel.py` inside of an `iqmc/` subfolder. Additionally, iQMC now calls it's own `loop_source`, `loop_partcile`, and `step_particle` functions. These changes are intended to reduce the complexity in the primary MCDC `loop` and `kernel` functions - most embedded iQMC conditional statements and loops have been removed (except for copying the `P["iqmc"]["w"]` because there is still only one particle type). This should also increase readability for both Monte Carlo and iQMC functions. 

2) The Scipy low-discrepency sequences have been removed and custom Halton sequences were added. `halton()` is the standard Halton sequence that is currently used for all iQMC simulations. `rhalton()` is the randomized-Halton sequence which will be used in a forth-coming batching method PR. This removes the dependence for `scipy`.

EDIT: 
I thought removing the halton sequence removed our dependence for `scipy`. However, we still use it in `input_.py` to integrate
```python
            # Make cdf
            card["energy"][1, :] = sp.integrate.cumulative_trapezoid(
                card["energy"][1], x=card["energy"][0], initial=0.0
            )

```

### CODE REVIEW

**Comment by jpmorgan98** [comment]
Really cool changes! IF/when we want to port iqmc to GPUs will this make that more difficult as we have more independent source files?

 I will run the GPU tests to ensure they still compile with all these changes and then merge

File: mcdc/iqmc/iqmc_kernel.py [hunk 1]
```diff
@@ -0,0 +1,714 @@
+#!/usr/bin/env python3
```
Comment (jpmorgan98) [inline mcdc/iqmc/iqmc_kernel.py:1]: Please remove these Spyder docstrings

File: mcdc/iqmc/iqmc_kernel.py [hunk 2]
```diff
@@ -0,0 +1,714 @@
...
+def scramble_LDS(mcdc):
+    # TODO: use MCDC seed system
+    seed_batch = np.random.randint(1000, 1000000)
+    N, dim = mcdc["technique"]["iqmc"]["lds"].shape
+    N_start = mcdc["mpi_work_start"]
+    mcdc["technique"]["iqmc"]["lds"] = rhalton(N, dim, seed=seed_batch, skip=N_start)
+
+
+@toggle("iQMC")
+def rhalton(N, dim, seed=12345, skip=0):
```
Comment (jpmorgan98) [inline mcdc/iqmc/iqmc_kernel.py:52]: Could you add some doc strings here? It doesn't need to be as robust as the input deck. The halton sequence seems to be undergriding iQMC so it seems mildly important to me but lmk

File: mcdc/iqmc/iqmc_loop.py [hunk 3]
```diff
@@ -0,0 +1,403 @@
+#!/usr/bin/env python3
```
Comment (jpmorgan98) [inline mcdc/iqmc/iqmc_loop.py:1]: Again please remove spyder doc strings

**Comment by jpmorgan98** [comment]
Also just so I understand it every iQMC function now uses the Halton sequence, right? Therefore it's no longer a togglable option with no need for `input_.py` documentation?

**Review by jpmorgan98 — CHANGES_REQUESTED** [review]
Just a few odds and ends mostly adding some spyder autodoc strings and adding some other seemingly important doc strings, if I get bored this weekend I will try and do a few of them myself. Thanks sam!!!

**Comment by ilhamv** [comment]
Thanks, Sam.Does the refactor solve the issue in the CSG PR?Sent from my iPhoneOn Jun 22, 2024, at 12:54 PM, Joanna Piper Morgan ***@***.***> wrote:﻿
@jpmorgan98 requested changes on this pull request.

Just a few odds and ends mostly adding some spyder autodoc strings and adding some other seemingly important doc strings, if I get bored this weekend I will try and do a few of them myself. Thanks sam!!!

In mcdc/iqmc/iqmc_kernel.py:
> @@ -0,0 +1,714 @@
+#!/usr/bin/env python3

Please remove these Spyder docstrings



In mcdc/iqmc/iqmc_kernel.py:
> +    N, dim = mcdc["technique"]["iqmc"]["lds"].shape
+    N_start = mcdc["mpi_work_start"]
+    mcdc["technique"]["iqmc"]["lds"] = halton(N, dim, skip=N_start)
+
+
***@***.***("iQMC")
+def scramble_LDS(mcdc):
+    # TODO: use MCDC seed system
+    seed_batch = np.random.randint(1000, 1000000)
+    N, dim = mcdc["technique"]["iqmc"]["lds"].shape
+    N_start = mcdc["mpi_work_start"]
+    mcdc["technique"]["iqmc"]["lds"] = rhalton(N, dim, seed=seed_batch, skip=N_start)
+
+
***@***.***("iQMC")
+def rhalton(N, dim, seed=12345, skip=0):

Could you add some doc strings here? It doesn't need to be as robust as the input deck. The halton sequence seems to be undergriding iQMC so it seems mildly important to me but lmk



In mcdc/iqmc/iqmc_loop.py:
> @@ -0,0 +1,403 @@
+#!/usr/bin/env python3

Again please remove spyder doc strings

**Comment by jpmorgan98** [comment]
Passed GPU tests

**Pull request merged**